### PR TITLE
Seedbox: download and upload torrent.

### PR DESCRIPTION
**Pull request by datnguyencse**
Add a seedbox page to handle torrent download, upload.

Screenshot from 2020-04-01 11-42-23

### CODE REVIEW

**Comment by team-pct** [comment]
Can one of the admins verify this patch?

File: src/app/lib/streamer.js [hunk 1]
```diff
@@ -63,20 +82,49 @@
...
+        },
+
         // kill the streamer
         stop: function() {
             if (this.torrent) {
                 // update ratio
                 AdvSettings.set('totalDownloaded', Settings.totalDownloaded + this.downloaded);
                 AdvSettings.set('totalUploaded', Settings.totalUploaded + this.uploaded);
-                this.torrent.destroy();
+                this.torrent.pause();
```
Comment (team-pct) [inline src/app/lib/streamer.js:119]: This is the same as torrent.destory()

File: src/app/settings.js [hunk 2]
```diff
@@ -176,6 +176,7 @@ Settings.databaseLocation = path.join(data_path, "data");
 Settings.deleteTmpOnClose = true;
 Settings.continueSeedingOnStart = false;
 Settings.vpnEnabled = true;
+Settings.maxActiveTorrents = 5;
```
Comment (team-pct) [inline src/app/settings.js:179]: Unable to Find that on Setting_container to modify this value

File: src/app/lib/device/ext-player.js [hunk 3]
```diff
@@ -221,12 +221,11 @@
...
         win.info('Scanning: ' + folderName);
         var appIndex = -1;
-        var fileStream = readdirp({
-            root: folderName,
+        var fileStream = readdirp(folderName, {
             depth: 3
         });
         fileStream.on('data', function (d) {
-            var app = d.name.replace('.app', '').replace('.exe', '').toLowerCase();
+            var app = d.basename.replace('.app', '').replace('.exe', '').toLowerCase();
```
Comment (team-pct) [inline src/app/lib/device/ext-player.js]: Uncaught TypeError: Cannot create property 'root' on string '/usr/bin' 
Error on OSX

File: src/app/lib/views/seedbox.js [hunk 4]
```diff
@@ -0,0 +1,258 @@
+(function (App) {
```
Comment (team-pct) [inline src/app/lib/views/seedbox.js:1]: User should be able to call the player from the Seedbox tab

File: src/app/templates/browser/filter-bar.tpl [hunk 5]
```diff
@@ -90,6 +90,11 @@
         <i id="filterbar-about" class="fa fa-info-circle about tooltipped" data-toggle="tooltip" data-placement="bottom" title="<%= i18n.__("About") %>"></i>
     </li>
 
+    <!-- Seedbox -->
```
Comment (team-pct) [inline src/app/templates/browser/filter-bar.tpl:93]: You should add Seedbox active Icon

**Review by team-pct — CHANGES_REQUESTED** [review]
Please look into review changes

**Comment by team-pct** [comment]
Tv Show are buggy with Seedbox please check again

**Comment by team-pct** [comment]
Capture d’écran 2020-04-01 à 15 43 29
Capture d’écran 2020-04-01 à 15 43 33

**Comment by team-pct** [comment]
The Selected TV Show should show Active

**Comment by ivan1986** [comment]
some fixes - https://github.com/popcorn-time-ru/popcorn-desktop/tree/seedbox

File: src/app/lib/device/ext-player.js (hunk 3 above)
Comment (ivan1986) [reply src/app/lib/device/ext-player.js]: reverted in my fork - link in comment

File: src/app/templates/browser/filter-bar.tpl (hunk 5 above)
Comment (ivan1986) [reply src/app/templates/browser/filter-bar.tpl:93]: added

**Review by ivan1986 — CHANGES_REQUESTED** [review]

File: src/app/lib/streamer.js (hunk 1 above)
Comment (ivan1986) [reply src/app/lib/streamer.js:119]: not, it's not same - destroy delete all files

File: src/app/database.js [hunk 6]
```diff
@@ -412,6 +412,10 @@ var Database = {
                 App.vent.trigger('initHttpApi');
                 App.vent.trigger('db:ready');
 
+                if (Settings.continueSeedingOnStart) {
+                  App.vent.trigger('stream:loadExistTorrents');
```
Comment (ivan1986) [inline src/app/database.js]: same if in start function - remove

File: src/app/lib/views/browser/filter_bar.js [hunk 7]
```diff
@@ -379,6 +401,8 @@
       App.currentview = "anime";
       App.vent.trigger("about:close");
       App.vent.trigger("torrentCollection:close");
+      App.vent.trigger("seedbox:close");
+      App.vent.trigger("seedbox:close");
```
Comment (ivan1986) [inline src/app/lib/views/browser/filter_bar.js]: double close - fixed

**Comment by ivan1986** [comment]
in serials torrent select after click to episode
i think easy way now is add download button near watch button and remove from list

**Comment by ivan1986** [comment]
https://github.com/datnguyencse/popcorn-desktop/pull/1

File: src/app/lib/streamer.js (hunk 1 above)
Comment (datnguyencse) [reply src/app/lib/streamer.js:119]: I wanna pause the torrent every time user close the player, they can resume or delete it from the download page. I'll change it back if you want.

Comment (team-pct) [reply src/app/lib/streamer.js:119]: No its fine please accept the PR from @ivan1986

File: src/app/lib/device/ext-player.js (hunk 3 above)
Comment (datnguyencse) [reply src/app/lib/device/ext-player.js]: I am using Linux and got error if `d.name` is used.
`app.js:36 [ERROR] Error [ERR_UNHANDLED_ERROR]: Unhandled error. (TypeError: Cannot read property 'replace' of undefined    at ReaddirpStream.<anonymous> (chrome-extension://hecfofbbdfadifpemejbbdcjmfmboohj/src/app/lib/device/ext-player.js:228:30)    at ReaddirpStream.emit (events.js:316:20)    at addChunk (_stream_readable.js:297:12)    at readableAddChunk (_stream_readable.js:273:9)    at ReaddirpStream.Readable.push (_stream_readable.js:214:10)    at ReaddirpStream._read (/home/seedceo/projects/popcorn-desktop/node_modules/readdirp/index.js:140:22)) Error [ERR_UNHANDLED_ERROR]: Unhandled error. (TypeError: Cannot read property 'replace' of undefined
    at ReaddirpStream.<anonymous> (chrome-extension://hecfofbbdfadifpemejbbdcjmfmboohj/src/app/lib/device/ext-player.js:228:30)
    at ReaddirpStream.emit (events.js:316:20)
    at addChunk (_stream_readable.js:297:12)
    at readableAddChunk (_stream_readable.js:273:9)
    at ReaddirpStream.Readable.push (_stream_readable.js:214:10)
    at ReaddirpStream._read (/home/seedceo/projects/popcorn-desktop/node_modules/readdirp/index.js:140:22))
    at ReaddirpStream.emit (events.js:305:17)
    at emitErrorCloseNT (internal/streams/destroy.js:75:8)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)`

Comment (team-pct) [reply src/app/lib/device/ext-player.js]: but it is supposed to look into OSX app

Comment (team-pct) [reply src/app/lib/device/ext-player.js]: Try the PR that was sent by @ivan1986

File: src/app/lib/views/player/loading.js [hunk 8]
```diff
@@ -187,7 +188,6 @@
       win.info("Loading torrent:", state);
 
       this.ui.stateTextDownload.text(i18n.__(state));
-      this.ui.stateTextStreamUrl.text(streamInfo.get('src').replace('127.0.0.1', Settings.ipAddress));
```
Comment (datnguyencse) [inline src/app/lib/views/player/loading.js:190]: I removed this line because every time I replay same movies, streamInfo.get('src') is undefined.

File: src/app/lib/device/ext-player.js (hunk 3 above)
Comment (datnguyencse) [reply src/app/lib/device/ext-player.js]: Because we're using different version of readdirp, latest version is v3.4.0 and there are no root option.

File: src/app/lib/streamer.js (hunk 1 above)
Comment (datnguyencse) [reply src/app/lib/streamer.js:119]: > No its fine please accept the PR from @ivan1986

Yep, I'm reviewing that PR

File: src/app/lib/views/seedbox.js (hunk 4 above)
Comment (datnguyencse) [reply src/app/lib/views/seedbox.js:1]: Should we add this play button to Seedbox tab?
Screenshot from 2020-04-01 11-07-04

**Comment by team-pct** [comment]
I am still have issue to download TV Show 
>Uncaught TypeError: Cannot read property 'magnet' of undefined
    at WebTorrentStreamer.download (streamer.js:98)
    at _ (backbone.js:369)
    at m (backbone.js:356)
    at f (backbone.js:155)
    at c.Channel.u.trigger (backbone.js:346)
    at n.downloadTorrent (show_detail.js:505)
    at N (underscore-min.js:5)
    at HTMLElement.<anonymous> (underscore-min.js:5)
    at HTMLElement.<anonymous> (underscore-min.js:5)
    at HTMLDivElement.dispatch (jquery.min.js:2)
database.js:264 Uncaught TypeError: Cannot read property 'toString' of undefined
    at Object.checkEpisodeWatched (database.js:264)
    at show_detail.js:330

**Comment by team-pct** [comment]
Also i am unable to Play directly from the Seedbox Tab

File: src/app/templates/show-detail.tpl [hunk 9]
```diff
@@ -75,6 +75,7 @@
                             </a>
 
                             <i id="watched-<%=episodeData.season%>-<%=episodeData.episode%>" class="fa fa-eye watched"></i>
+                            <i id="download-torrent" class="fa fa-download watched"></i>
```
Comment (ivan1986) [inline src/app/templates/show-detail.tpl]: better move this to overview tab for each episode (near watch button)
here we have duplicate id and need select quality for episode

**Review by ivan1986 — CHANGES_REQUESTED** [review]
https://github.com/datnguyencse/popcorn-desktop/pull/2

**Comment by datnguyencse** [comment]
> I am still have issue to download TV Show

Can you try again, I just merge ivan1986's PR.
Screenshot from 2020-04-02 21-40-14

File: src/app/settings.js (hunk 2 above)
Comment (datnguyencse) [reply src/app/settings.js:179]: added

File: src/app/templates/show-detail.tpl (hunk 9 above)
Comment (datnguyencse) [reply src/app/templates/show-detail.tpl]: Thanks for the PR.

**Comment by team-pct** [comment]
> > I am still have issue to download TV Show
> 
> Can you try again, I just merge ivan1986's PR.
> Screenshot from 2020-04-02 21-40-14

Is working now can you add the ability to stream the File directly from Seedbox tab?

**Comment by BGamboa13** [comment]
Changing cache directory, seedbox tab not working! @team-pct @datnguyencse

File: src/app/lib/views/seedbox.js [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+                <i class="fa fa-trash-alt watched trash-torrent" id="trash-${torrent.infoHash}"></i>
+                <i class="fa fa-play watched resume-torrent" id="play-${torrent.infoHash}" style="display: ${torrent.paused ? '' : 'none'};"></i>
+                <i class="fa fa-pause-circle watched pause-torrent" id="resume-${torrent.infoHash}" style="display: ${torrent.paused ? 'none' : ''};"></i>
+                <i class="fa fa-upload watched" id="upload-${torrent.infoHash}"> 0 Kb/s</i>
+                <i class="fa fa-download watched" id="download-${torrent.infoHash}"> 0 Kb/s</i>
+              </li>`
+            );
+
+            $('.seedbox-overview').show();
+          }, 500 + Math.floor(Math.random() * Math.floor(500)));
```
Comment (ArcticZeroo) [inline src/app/lib/views/seedbox.js:159]: What the heck is this here for? I'm going through the code for an unrelated change and this just seems wholly unnecessary.

Comment (datnguyencse) [reply src/app/lib/views/seedbox.js:159]: @ArcticZeroo 500+ milliseconds is necessary to make sure the torrent is fully initialized, as you can see in the first line for this function, I also check `if ($(`#${torrent.infoHash}`).length || !torrent.name)`. So what will happen if no delay in the function? => nothing will be add to the seedbox UI.

Comment (datnguyencse) [reply src/app/lib/views/seedbox.js:159]: @ArcticZeroo I'm not sure why I added the setTimeout here, but it must be for a reason. You're welcome to test, remove it and create a new merger request. Thank you.

**Comment by datnguyencse** [comment]
> Changing cache directory, seedbox tab not working! @team-pct @datnguyencse

Thank for the report, I'm going to try it myself and fix if it's a bug.

File: src/app/lib/views/seedbox.js (hunk 10 above)
Comment (ArcticZeroo) [reply src/app/lib/views/seedbox.js:159]: Sure I guess, but the solution to an object that takes time to initialize is not to wait 500ms + a random amount of milliseconds (seriously, how does that not seem crazy?), but rather to wait for it to be ready, e.g. `torrent.on('ready')`

Comment (datnguyencse) [reply src/app/lib/views/seedbox.js:159]: Thank, it's my bad. i wrote the code for my daily use only - and didn't spend more time to optimize, then I realized It's useful and make a new merge request without any change. Please help to improve it.

Comment (ArcticZeroo) [reply src/app/lib/views/seedbox.js:159]: it probably needs more cleanup eventually, but this particular change is in https://github.com/popcorn-official/popcorn-desktop/pull/1560
